src/components/vm/vmActions.tsx

```diff
@@ -44,44 +44,88 @@ import { RenameDialog } from './vmRenameDialog.jsx';
-    domainCanRename,
-    domainCanResume,
-    domainCanRun,
-    domainCanPause,
-    domainCanShutdown,
-    domainForceOff,
-    domainForceReboot,
+    domainCanInstall,
     domainInstall,
-    domainPause,
-    domainReboot,
-    domainResume,
-    domainSendNMI,
-    domainShutdown,
-    domainStart,
+    vmDomainMethod,
     virtXmlAdd
 } from '../../libvirtApi/domain.js';
 import store from "../../store.js";
 
 const _ = cockpit.gettext;
 
-const onStart = (vm: VM, setOperationInProgress: (val: boolean) => void) => (
-    domainStart({ id: vm.id, connectionName: vm.connectionName }).catch(ex => {
-        setOperationInProgress(false);
-        console.warn("Failed to start VM", vm.name, ":", cockpit.message(ex));
-        store.dispatch(
-            updateVm({
-                connectionName: vm.connectionName,
-                name: vm.name,
-                error: {
-                    text: cockpit.format(_("VM $0 failed to start"), vm.name),
-                    detail: ex.message,
-                }
-            })
-        );
-    })
-);
+const domainCanRun = (vmState: VMState, hasInstallPhase: boolean) => !hasInstallPhase && vmState == 'shut off';
+const domainCanShutdown = (vmState: VMState) => domainCanReset(vmState);
+const domainCanPause = (vmState: VMState) => vmState == 'running';
+const domainCanRename = (vmState: VMState) => vmState == 'shut off';
+const domainCanResume = (vmState: VMState) => vmState == 'paused';
+
+function startOperationProgress(vm: VM, from_state: VMState = vm.state) {
+    store.dispatch(
+        updateVm({
+            connectionName: vm.connectionName,
+            name: vm.name,
+            operationInProgressFromState: from_state,
+            onShutOff: null,
+        })
+    );
+}
+
+function resetOperationProgress(vm: VM) {
+    store.dispatch(
+        updateVm({
+            connectionName: vm.connectionName,
+            name: vm.name,
+            operationInProgressFromState: undefined,
+            onShutOff: null,
+        })
+    );
+}
+
+function isOperationInProgress(vm: VM) {
+    return vm.state == vm.operationInProgressFromState;
+}
+
+function setVmError(vm: VM, msg: string, ex: unknown) {
+    console.warn(msg, ":", String(ex));
+    store.dispatch(
+        updateVm({
+            connectionName: vm.connectionName,
+            name: vm.name,
+            error: {
+                text: msg,
+                detail: String(ex),
+            }
+        })
+    );
+}
+
+async function actionWithProgress(vm: VM, action: () => Promise<void>, msg: string): Promise<void> {
+    startOperationProgress(vm);
+    try {
+        await action();
+    } catch (ex) {
+        resetOperationProgress(vm);
+        setVmError(vm, msg, ex);
+    }
+}
+
+async function actionWithoutProgress(vm: VM, action: () => Promise<void>, msg: string): Promise<void> {
+    resetOperationProgress(vm);
+    try {
+        await action();
+    } catch (ex) {
+        setVmError(vm, msg, ex);
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2376-a00f515a-20251027-132245-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)